include/pybind11/pytypes.h

```diff
@@ -1784,24 +1784,41 @@ class kwargs : public dict {
     PYBIND11_OBJECT_DEFAULT(kwargs, dict, PyDict_Check)
 };
 
-class set : public object {
+class any_set : public object {
+protected:
+    PYBIND11_OBJECT(any_set, object, PyAnySet_Check)
+
+public:
+    size_t size() const { return (size_t) PySet_Size(m_ptr); }
+    bool empty() const { return size() == 0; }
+    template <typename T>
+    bool contains(T &&val) const {
+        return PySet_Contains(m_ptr, detail::object_or_cast(std::forward<T>(val)).ptr()) == 1;
+    }
+};
+
+class set : public any_set {
 public:
-    PYBIND11_OBJECT_CVT(set, object, PySet_Check, PySet_New)
-    set() : object(PySet_New(nullptr), stolen_t{}) {
+    PYBIND11_OBJECT_CVT(set, any_set, PySet_Check, PySet_New)
+    set() : any_set(PySet_New(nullptr), stolen_t{}) {
         if (!m_ptr) {
             pybind11_fail("Could not allocate set object!");
```

Comment:
Just a remark, not actionable in this PR: `throw error_already_set()` would seem better, similar to the changes @Skylion007 made elsewhere recently. But strictly speaking that's a behavior change, therefore clearly better left for a separate PR IMO.

Reply:
@rwgk No, this is correct to my knowledge. `PySet_New` only throws one type of an exception: a TypeError if the sequence fed into it is not a valid type. The sequence fed into it is a nullptr, so it won't raise that exception. It will return NULL if another unspecified error occurs, but it won't raise an exception to my knowledge: https://docs.python.org/3/c-api/set.html . 
If we add a C++ constructor that takes in an iterable though and passes it into the arg for `PySet_New`, then we should add the `throw error_already_set` in that ctor though.